File: src/Core/OntologyAnnotation.fs
namespace ARCtrl

open ARCtrl.Helper
open Fable.Core

[<AttachMembers>]
// tanInfo given from outside was meant to allow for performance optimization, but it is not used in the code currently.
type OntologyAnnotation(?name,?tsr,?tan, (*?tanInfo, *)?comments) =
    let mutable _name : string option =
        match name with
        | Some "" | None -> None
        | name -> name
    let mutable _termSourceREF : string option =
        match tsr with
        | Some "" | None -> None
        | tsr -> tsr
    let mutable _termAccessionNumber : string option =
        match tan with
        | Some "" | Some ":" | None -> None
        | tan -> tan
    let mutable _comments : ResizeArray<Comment> = defaultArg comments <| ResizeArray()

    let mutable _tanInfo : {| IDSpace : string; LocalID : string |} option =
        //match tanInfo with
        //| Some tanInfo -> tanInfo
        //| None ->
        OntologyAnnotation.computeTanInfo (?tan = _termAccessionNumber, ?tsr = _termSourceREF)

    let recomputeTanInfo() =
        let newTanInfo = OntologyAnnotation.computeTanInfo(?tan = _termAccessionNumber, ?tsr = _termSourceREF)
        _tanInfo <- newTanInfo

    member this.Name
        with get() = _name
        and set(name) = _name <- name

    member this.TermSourceREF
        with get() = _termSourceREF
        and set(tsr) =
            _termSourceREF <- tsr
            recomputeTanInfo()

    member this.TermAccessionNumber
        with get() = _termAccessionNumber
        and set(tan) =
            _termAccessionNumber <- tan
            recomputeTanInfo()

    member this.Comments
        with get() = _comments
        and set(comments) = _comments <- comments

    static member make name tsr tan comments = OntologyAnnotation(?name=name, ?tsr=tsr, ?tan=tan, comments=comments)
    
    ///<summary>
    /// Create a ISAJson Ontology Annotation value from ISATab string entries, will try to reduce `termAccessionNumber` with regex matching.
    ///
    /// Exmp. 1: http://purl.obolibrary.org/obo/GO_000001 --> GO:000001
    ///</summary>
    ///<param name="name"></param>
    ///<param name="tsr">Term accession number</param>
    ///<param name="tan">Term accession number</param>
    ///<param name="comments">Term accession number</param>
    static member create(?name,?tsr,?tan, ?comments) : OntologyAnnotation =
        OntologyAnnotation.make name tsr tan (defaultArg comments <| ResizeArray())

    member this.TANInfo = _tanInfo

    member this.NameText = 
        this.Name
        |> Option.defaultValue ""

    /// Create a path in form of `http://purl.obolibrary.org/obo/MS_1000121` from it's Term Accession Source `MS` and Local Term Accession Number `1000121`. 
    static member createUriAnnotation (termSourceRef : string) (localTAN : string) =
        Helper.Url.createOAUri termSourceRef localTAN

    static member fromTermAnnotation (tan : string, ?name) =
        match Regex.tryParseTermAnnotation tan with
        | Some r ->
            let accession = r.IDSpace + ":" + r.LocalID
            OntologyAnnotation.create (?name = name, tsr = r.IDSpace, tan = accession)
        | None -> OntologyAnnotation.create (?name = name, tan = tan)

    /// Parses any value in `TermAccessionString` to term accession format "termsourceref:localtan". Exmp.: "MS:000001".
    ///
    /// If `TermAccessionString` cannot be parsed to this format, returns empty string!
    member this.TermAccessionShort = 
        match this.TANInfo with
        | Some id -> $"{id.IDSpace}:{id.LocalID}"
        | _ -> ""

    member this.TermAccessionOntobeeUrl = 
        match this.TANInfo with
        | Some id -> OntologyAnnotation.createUriAnnotation id.IDSpace id.LocalID
        | _ -> ""

    member this.TermAccessionAndOntobeeUrlIfShort = 
        match this.TermAccessionNumber with
        | Some tan -> 
            match tan with 
            | Regex.ActivePatterns.Regex Regex.Pattern.TermAnnotationShortPattern _ -> this.TermAccessionOntobeeUrl
            | _ -> tan
        | _ -> ""

    /// <summary>
    /// Get a ISATab string entries from an ISAJson Ontology Annotation object (name,source,accession)
    ///
    /// `asOntobeePurlUrl`: option to return term accession in Ontobee purl-url format (`http://purl.obolibrary.org/obo/MS_1000121`)
    /// </summary>
    static member toStringObject (oa : OntologyAnnotation, ?asOntobeePurlUrlIfShort: bool) =
        let asOntobeePurlUrlIfShort = Option.defaultValue false asOntobeePurlUrlIfShort
        {|
            TermName = oa.Name |> Option.defaultValue ""
            TermSourceREF = oa.TermSourceREF |> Option.defaultValue ""
            TermAccessionNumber = 
                if asOntobeePurlUrlIfShort then
                    let url = oa.TermAccessionAndOntobeeUrlIfShort
                    if url = "" then 
                        oa.TermAccessionNumber |> Option.defaultValue ""
                    else
                        url
                else
                    oa.TermAccessionNumber |> Option.defaultValue ""
        |}

    interface IISAPrintable with
        member this.Print() =
            this.ToString()
        member this.PrintCompact() =
            "OA " + this.NameText

    override this.ToString() =
        let sb = System.Text.StringBuilder()
        sb.Append("{") |> ignore
        [
            if this.Name.IsSome then
                sprintf "Name = %A" this.Name.Value
            if this.TermSourceREF.IsSome then
                sprintf "TSR = %A" this.TermSourceREF.Value
            if this.TermAccessionNumber.IsSome then
                sprintf "TAN = %A" this.TermAccessionNumber.Value
            if this.Comments.Count <> 0 then
                sprintf "Comments = %A" this.Comments
        ] 
        |> String.concat "; "
        |> sb.Append
        |> ignore
        sb.Append("}") |> ignore
        sb.ToString()

    member this.isEmpty() = 
           this.Name.IsNone && this.TermSourceREF.IsNone && this.TermAccessionNumber.IsNone && this.Comments.Count = 0

    override this.GetHashCode() = 
        // Same as above but don't create unneccessary arrays
        let nameHash = 
            match this.Name with
            | Some name -> hash name
            | None -> 0
        match this.TermSourceREF, this.TANInfo with
        | None, Some taninfo -> 
            HashCodes.mergeHashes (hash taninfo.IDSpace) (hash taninfo.LocalID)
        | Some tsr, Some taninfo -> 
            HashCodes.mergeHashes (hash tsr) (hash taninfo.LocalID)
        | Some tsr, None ->
            match this.TermAccessionNumber with
            | Some tan -> HashCodes.mergeHashes (hash tsr) (hash tan)
            | None -> hash tsr
        | None, None ->
            0    
        |> HashCodes.mergeHashes nameHash

    override this.Equals(obj) : bool =
        HashCodes.hash this = HashCodes.hash obj
   
    interface System.IComparable with
        member this.CompareTo(obj) =
            match obj with
            | :? OntologyAnnotation as oa -> 
                #if FABLE_COMPILER
                let hash = this.GetHashCode()
                let otherHash = oa.GetHashCode()
                if hash = otherHash then 0
                else if hash < otherHash then -1
                else 1                
                #else
                this.GetHashCode().CompareTo(oa.GetHashCode())
                #endif
            | _ -> 1

    member this.Copy() =
        let nextComments = this.Comments |> ResizeArray.map (fun c -> c.Copy())
        OntologyAnnotation.make this.Name this.TermSourceREF this.TermAccessionNumber nextComments

    static member computeTanInfo (?tan : string, ?tsr : string) =
        match tan with
        | Some tan -> 
            match Regex.tryParseTermAnnotation tan with
            | Some ta -> Some ta
            | None ->
                match tsr with
                | Some "" | None -> None
                | Some tsr -> 
                    Some {| IDSpace = tsr; LocalID = tan |}
        | None -> None

File: js/Core/OntologyAnnotation.js
import { value as value_4, defaultArg, unwrap } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { Pattern_TermAnnotationShortPattern, ActivePatterns_$007CRegex$007C_$007C, tryParseTermAnnotation } from "./Helper/Regex.js";
import { createOAUri } from "./Helper/Url.js";
import { StringBuilder__Append_Z721C83C5, StringBuilder_$ctor } from "../fable_modules/fable-library-js.4.22.0/System.Text.js";
import { printf, toText, join } from "../fable_modules/fable-library-js.4.22.0/String.js";
import { toArray, empty, singleton, append, delay, toList } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { toString } from "../fable_modules/fable-library-js.4.22.0/Types.js";
import { hash as hash_1, boxHashOption, boxHashArray } from "./Helper/HashCodes.js";
import { ResizeArray_map } from "./Helper/Collections.js";
import { safeHash } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { class_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class OntologyAnnotation {
    constructor(name, tsr, tan, comments) {
        this._name = ((name == null) ? undefined : ((name === "") ? undefined : name));
        this._termSourceREF = ((tsr == null) ? undefined : ((tsr === "") ? undefined : tsr));
        this._termAccessionNumber = ((tan == null) ? undefined : ((tan === "") ? undefined : ((tan === ":") ? undefined : tan)));
        this._comments = defaultArg(comments, []);
    }
    get Name() {
        const this$ = this;
        return unwrap(this$._name);
    }
    set Name(name) {
        const this$ = this;
        this$._name = name;
    }
    get TermSourceREF() {
        const this$ = this;
        return unwrap(this$._termSourceREF);
    }
    set TermSourceREF(tsr) {
        const this$ = this;
        this$._termSourceREF = tsr;
    }
    get TermAccessionNumber() {
        const this$ = this;
        return unwrap(this$._termAccessionNumber);
    }
    set TermAccessionNumber(tan) {
        const this$ = this;
        this$._termAccessionNumber = tan;
    }
    get Comments() {
        const this$ = this;
        return this$._comments;
    }
    set Comments(comments) {
        const this$ = this;
        this$._comments = comments;
    }
    static make(name, tsr, tan, comments) {
        return new OntologyAnnotation(unwrap(name), unwrap(tsr), unwrap(tan), comments);
    }
    static create(name, tsr, tan, comments) {
        const comments_1 = defaultArg(comments, []);
        return OntologyAnnotation.make(name, tsr, tan, comments_1);
    }
    get TANInfo() {
        let matchValue, tan, matchValue_1, matchValue_2, tsr;
        const this$ = this;
        return unwrap((matchValue = this$.TermAccessionNumber, (matchValue == null) ? undefined : ((tan = matchValue, (matchValue_1 = tryParseTermAnnotation(tan), (matchValue_1 == null) ? ((matchValue_2 = this$.TermSourceREF, (matchValue_2 == null) ? undefined : ((matchValue_2 === "") ? undefined : ((tsr = matchValue_2, {
            IDSpace: tsr,
            LocalID: tan,
        }))))) : matchValue_1)))));
    }
    get NameText() {
        const this$ = this;
        return defaultArg(this$.Name, "");
    }
    static createUriAnnotation(termSourceRef, localTAN) {
        return createOAUri(termSourceRef, localTAN);
    }
    static fromTermAnnotation(tan, name) {
        const matchValue = tryParseTermAnnotation(tan);
        if (matchValue == null) {
            return OntologyAnnotation.create(unwrap(name), undefined, tan);
        }
        else {
            const r = matchValue;
            const accession = (r.IDSpace + ":") + r.LocalID;
            return OntologyAnnotation.create(unwrap(name), r.IDSpace, accession);
        }
    }
    get TermAccessionShort() {
        const this$ = this;
        const matchValue = this$.TANInfo;
        if (matchValue != null) {
            const id = matchValue;
            return `${id.IDSpace}:${id.LocalID}`;
        }
        else {
            return "";
        }
    }
    get TermAccessionOntobeeUrl() {
        const this$ = this;
        const matchValue = this$.TANInfo;
        if (matchValue != null) {
            const id = matchValue;
            return OntologyAnnotation.createUriAnnotation(id.IDSpace, id.LocalID);
        }
        else {
            return "";
        }
    }
    get TermAccessionAndOntobeeUrlIfShort() {
        const this$ = this;
        const matchValue = this$.TermAccessionNumber;
        if (matchValue != null) {
            const tan = matchValue;
            return (ActivePatterns_$007CRegex$007C_$007C(Pattern_TermAnnotationShortPattern, tan) != null) ? this$.TermAccessionOntobeeUrl : tan;
        }
        else {
            return "";
        }
    }
    static toStringObject(oa, asOntobeePurlUrlIfShort) {
        let url;
        const asOntobeePurlUrlIfShort_1 = defaultArg(asOntobeePurlUrlIfShort, false);
        const TermName = defaultArg(oa.Name, "");
        const TermSourceREF = defaultArg(oa.TermSourceREF, "");
        return {
            TermAccessionNumber: asOntobeePurlUrlIfShort_1 ? ((url = oa.TermAccessionAndOntobeeUrlIfShort, (url === "") ? defaultArg(oa.TermAccessionNumber, "") : url)) : defaultArg(oa.TermAccessionNumber, ""),
            TermName: TermName,
            TermSourceREF: TermSourceREF,
        };
    }
    toString() {
        const this$ = this;
        const sb = StringBuilder_$ctor();
        StringBuilder__Append_Z721C83C5(sb, "{");
        StringBuilder__Append_Z721C83C5(sb, join("; ", toList(delay(() => {
            let arg;
            return append((this$.Name != null) ? singleton((arg = value_4(this$.Name), toText(printf("Name = %A"))(arg))) : empty(), delay(() => {
                let arg_1;
                return append((this$.TermSourceREF != null) ? singleton((arg_1 = value_4(this$.TermSourceREF), toText(printf("TSR = %A"))(arg_1))) : empty(), delay(() => {
                    let arg_2;
                    return append((this$.TermAccessionNumber != null) ? singleton((arg_2 = value_4(this$.TermAccessionNumber), toText(printf("TAN = %A"))(arg_2))) : empty(), delay(() => {
                        let arg_3;
                        return (this$.Comments.length !== 0) ? singleton((arg_3 = this$.Comments, toText(printf("Comments = %A"))(arg_3))) : empty();
                    }));
                }));
            }));
        }))));
        StringBuilder__Append_Z721C83C5(sb, "}");
        return toString(sb);
    }
    isEmpty() {
        const this$ = this;
        return (((this$.Name == null) && (this$.TermSourceREF == null)) && (this$.TermAccessionNumber == null)) && (this$.Comments.length === 0);
    }
    GetHashCode() {
        const this$ = this;
        return boxHashArray(toArray(delay(() => append(singleton(boxHashOption(this$.Name)), delay(() => {
            let tsr_1, taninfo_1, tsr;
            const matchValue = this$.TermSourceREF;
            const matchValue_1 = this$.TANInfo;
            if (matchValue != null) {
                return (matchValue_1 == null) ? ((tsr_1 = matchValue, singleton(boxHashArray([tsr_1, defaultArg(this$.TermAccessionNumber, "")])))) : ((taninfo_1 = matchValue_1, (tsr = matchValue, singleton(boxHashArray([tsr, (taninfo_1.IDSpace + ":") + taninfo_1.LocalID])))));
            }
            else if (matchValue_1 == null) {
                const tan_1 = defaultArg(this$.TermAccessionNumber, "");
                return singleton(boxHashArray([defaultArg(this$.TermAccessionNumber, ""), tan_1]));
            }
            else {
                const taninfo = matchValue_1;
                return singleton(boxHashArray([taninfo.IDSpace, (taninfo.IDSpace + ":") + taninfo.LocalID]));
            }
        }))))) | 0;
    }
    Equals(obj) {
        const this$ = this;
        return hash_1(this$) === hash_1(obj);
    }
    Copy() {
        const this$ = this;
        const nextComments = ResizeArray_map((c) => c.Copy(), this$.Comments);
        const name = this$.Name;
        const tsr = this$.TermSourceREF;
        const tan = this$.TermAccessionNumber;
        return OntologyAnnotation.make(name, tsr, tan, nextComments);
    }
    Print() {
        const this$ = this;
        return toString(this$);
    }
    PrintCompact() {
        const this$ = this;
        return "OA " + this$.NameText;
    }
    CompareTo(obj) {
        const this$ = this;
        if (obj instanceof OntologyAnnotation) {
            const oa = obj;
            const hash = safeHash(this$) | 0;
            const otherHash = safeHash(oa) | 0;
            return ((hash === otherHash) ? 0 : ((hash < otherHash) ? -1 : 1)) | 0;
        }
        else {
            return 1;
        }
    }
}

export function OntologyAnnotation_$reflection() {
    return class_type("ARCtrl.OntologyAnnotation", undefined, OntologyAnnotation);
}

export function OntologyAnnotation_$ctor_Z54349580(name, tsr, tan, comments) {
    return new OntologyAnnotation(name, tsr, tan, comments);
}

